sqlglot/optimizer/scope.py

```diff
@@ -548,30 +549,39 @@ def _traverse_select(scope):
 
 
 def _traverse_union(scope):
-    yield from _traverse_ctes(scope)
+    prev_scope = None
+    union_scope_stack = [scope]
+    expression_stack = [scope.expression.left, scope.expression.right]
 
-    # The last scope to be yield should be the top most scope
-    left = None
-    for left in _traverse_scope(
-        scope.branch(
-            scope.expression.left,
-            outer_column_list=scope.outer_column_list,
-            scope_type=ScopeType.UNION,
-        )
-    ):
-        yield left
-
-    right = None
-    for right in _traverse_scope(
-        scope.branch(
-            scope.expression.right,
-            outer_column_list=scope.outer_column_list,
-            scope_type=ScopeType.UNION,
-        )
-    ):
-        yield right
+    while expression_stack:
+        expression = expression_stack.pop()
 
-    scope.union_scopes = [left, right]
+        if isinstance(expression, exp.Union):
+            union_scope = Scope(expression, parent=seq_get(union_scope_stack, -1))
```

Comment:
how come you're not using branch?

Reply:
I changed this locally, using branch now. This was probably the reason why the tests failed.